### PR TITLE
Added new functions: TrimDomain(), AddDomain(), GenerateZoneFile()

### PR DESCRIPTION
**Pull request by TomOnTime**
* TrimDomain() Trims the domain off a FQDN.
  * AddDomain() Adds a domain to a shortname.
  * GenerateZoneFile() Create a pretty BIND-style Zonefile.

### CODE REVIEW

File: zone.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+	if origin[len(origin)-1] == '.' {
+		suffix = origin[:len(origin)-1]
+	}
+	if name == suffix {
+		return "@"
+	}
+	return strings.TrimSuffix(name, "."+suffix)
+}
+
+type Zone struct {
```
Comment (miekg) [inline zone.go]: I once added a similar zone structure to ldns: http://git.nlnetlabs.nl/ldns/tree/ldns/zone.h?h=release-1.6.17#n42 This was also a simple list; this of course has scaling issues. Another thing it that most users of this lib have wildly different ideas on how to implement the concept of a Zone, that's one of the reason there isn't a zone structure in go dns.

File: zone.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+	"bytes"
+	"fmt"
+	"io"
+	"log"
+	"sort"
+	"strings"
+)
+
+// AddDomain adds origin to s if s is not already a FQDN.
+func AddDomain(origin, s string) string {
```
Comment (miekg) [inline zone.go]: I can see the benefit of these helpers. But take for instance `IsEdns0` in `default.go` - which I added at the time, because I thought: will be useful for some. People just code this out themselves. The a) don't expect such a function to be present or b) don't find it in the docs. In retrospect I think most functions in `default.go` should not have been added. I'm sorry, but I feel the same about these.

File: zone.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+	// If you want the result to end in a ".", pass an origin that ends in a "."
+	// ("origin.", "foo") -> "foo.origin."
+	// ("origin", "foo") -> "foo.origin"
+	// ("origin.", "foo.") -> "foo." (already a FQDN)
+	// ("origin.", "@") -> "origin." (@ represents the apex domain)
+	// ("origin.", "") -> "origin." (not obvious)
+	if len(origin) == 0 || origin[0] == '.' {
+		log.Fatalf("Invalid origin: %#v\n", origin)
```
Comment (miekg) [inline zone.go]: Always try to return an error instead of Fatalf-ing

File: zone.go (hunk 2 above)
Comment (TomOnTime) [inline zone.go]: Logically one can't know if utility functions like IsEdns0, AddDomain, TrimDomain are used by project outside of this module.  I agree that documenting them would help people find them.  I know that getting started using the module was difficult for me because the documentation was so brief.

Also, while a function such as IsEdns0() may seem obvious to some, the minutiae of DNS is staggeringly huge. Such utility functions prevent stupid bugs from haunting the internet.  For example without CountLabel, most people would just use strings.Count(s, "."), without IsSubDomain most people would just use strings.IsSubDomain(s, domain).

That said...

It sounds to me like this project only wants contributions in a very specific domain of functionality.  I'm not sure that that domain is, but I think it would make the project easier to contribute to if that domain was described in the README.md.  Feel free to contact me off-line if you'd like to know what I think it is, based on reverse engineering the comments posted on this PR.

File: zone.go (hunk 1 above)
Comment (TomOnTime) [inline zone.go]: It only needs a struct so that it can be sorted. I could create a simple `type sortableRR []RR` type just to create an interface that matches Sort.  I could re-factor the code to not need Origin and DefaultTtl by assuming the RRs have FQDNs and proper TTL set already.

File: zone.go (hunk 2 above)
Comment (miekg) [inline zone.go]: [ Quoting <notifications@github.com> in "Re: [dns] Added new functions: Trim..." ]
>Logically one can't know if utility functions like IsEdns0, AddDomain, TrimDomain are used by project outside of this module.  I agree that documenting them would help people find them.  I know that getting started using the module was difficult for me because the documentation was so brief.
>
>Also, while a function such as IsEdns0() may seem obvious to some, the minutiae of DNS is staggeringly huge. Such utility functions prevent stupid bugs from haunting the internet.  For example without CountLabel, most people would just use strings.Count(s, "."), without IsSubDomain most people would just use strings.IsSubDomain(s, domain).
>
>That said...
>
>It sounds to me like this project only wants contributions in a very specific domain of functionality.  I'm not sure that that domain is, but I think it would make the project easier to contribute to if that domain was described in the README.md.  Feel free to contact me off-line if you'd like to know what I think it is, based on reverse engineering the comments posted on this PR.

Fair enough. Having a more offline discussion makes sense. And I of course
welcome contributions, but I also don't want to bloat the API. This goes back
to the line (that is in the README :) ):

  If there is stuff you should know as a DNS programmer there isn't a convenience
  function for it.

But that is of course really open to all kinds of interpretation(s).

I don't want to alienate people by not accepting PRs and I don't want to make Go
DNS a "everything and the kitchen sink" kind of package. So it's a balance.

Happy to discuss this some more.

/Miek

--
Miek Gieben

Comment (asergeyev) [inline zone.go]: Could it be similar to IDN when things live within this library but in separate import path?

Comment (miekg) [inline zone.go]: [ Quoting <notifications@github.com> in "Re: [dns] Added new functions: Trim..." ]
>> @@ -0,0 +1,244 @@
>> +package dns
>> +
>> +import (
>> +	"bytes"
>> +	"fmt"
>> +	"io"
>> +	"log"
>> +	"sort"
>> +	"strings"
>> +)
>> +
>> +// AddDomain adds origin to s if s is not already a FQDN.
>> +func AddDomain(origin, s string) string {
>
>Could it be similar to IDN when things live within this library but in separate import path?

Hmmm, hadn't considered that. Nice idea.

/Miek

--
Miek Gieben

**Comment by miekg** [comment]
Lets go with that. Add a `util` dir and put TrimDomain() and AddDomain() (still not sure about GenerateZoneFile) in there.

**Comment by andrewtj** [comment]
> On 21 Jan 2016, at 8:21 AM, Miek Gieben <notifications@github.com> wrote:
> 
> Lets go with that. Add a util dir and put TrimDomain() and AddDomain() (still not sure about GenerateZoneFile) in there.
> 

`io` packages utility functions in `io/ioutil`, should we follow that convention? Might avoid some ambiguity.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [dns] Added new functions: Trim..." ]
>
>> On 21 Jan 2016, at 8:21 AM, Miek Gieben <notifications@github.com> wrote:
>>
>> Lets go with that. Add a util dir and put TrimDomain() and AddDomain() (still not sure about GenerateZoneFile) in there.
>>
>
>`io` packages utility functions in `io/ioutil`, should we follow that convention? Might avoid some ambiguity.

yep, like that suggestion too, dns/dnsutil it is.

Thanks!

**Comment by miekg** [comment]
@TomOnTime you agree to this as well? And if so, are you willing to rework this a little?

**Comment by TomOnTime** [comment]
Oh good!  I was going to suggest dnsutil too!   I'll try to get to this over the weekend.

**Comment by TomOnTime** [comment]
PTAL

File: dnsutil/util.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
+package dnsutil
+
+import "strings"
+
+// AddDomain adds origin to s if s is not already a FQDN.
+func AddDomain(origin, s string) string {
```
Comment (miekg) [inline dnsutil/util.go]: To be in sync with the names of the functions in `labels.go`, this is better named `AddDomainName` or maybe `AddOrigin`  and then reverse the order of the parameters.

File: dnsutil/util.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	// If you want the result to end in a ".", pass an origin that ends in a "."
```
Comment (miekg) [inline dnsutil/util.go]: Should we return an error on a non-qualified origin?

File: dnsutil/util.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	// ("origin.", "foo") -> "foo.origin."
+	// ("origin", "foo") -> "foo.origin"
+	// ("origin.", "foo.") -> "foo." (already a FQDN)
+	// ("origin.", "@") -> "origin." (@ represents the apex domain)
```
Comment (miekg) [inline dnsutil/util.go]: That `@` is specific should be moved to the function's doc.

File: dnsutil/util.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+		return origin
+	}
+	if origin == "." {
+		if s[len(s)-1] == '.' {
+			return s
+		} else {
+			return s + "."
+		}
+	}
+	if s == "*" {
```
Comment (miekg) [inline dnsutil/util.go]: why is this special cased?

File: dnsutil/util.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+		return "*." + origin
+	}
+	if s[len(s)-1] == '.' {
+		return s
+	}
+	return s + "." + origin
+}
+
+// TrimDomain trims the origin from a FQDN if it exists.
+func TrimDomain(origin, s string) string {
```
Comment (miekg) [inline dnsutil/util.go]: TrimDomainName (also believe it's more natural to flip the order of the arguments)

File: dnsutil/util.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+	// The apex domain is always returned as "@".
```
Comment (miekg) [inline dnsutil/util.go]: Should be moved to the function's doc. Not sure if `apex` is directly understood by every consumer of this function

File: dnsutil/util.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
+package dnsutil
```
Comment (miekg) [inline dnsutil/util.go]: package doc?

File: dnsutil/util.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	// We're being asked to trim something that is empty, so we special-case:
+	if len(s) == 0 {
+		return "@"
+	}
+	if s == "." {
+		return "."
+	}
+
+	lod := origin[len(origin)-1] == '.' // Is the last char of origin a dot?
```
Comment (miekg) [inline dnsutil/util.go]: `IsFqn(origin)` - maybe we should copy to this package as well btw.

File: dnsutil/util.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+	lsd := s[len(s)-1] == '.'           // Is the last char of s a dot?
+
+	if lod == lsd { // Either they both end in ".", or neither do.
+		if origin == s {
+			return "@"
+		}
+		if origin[0] == '.' {
+			return strings.TrimSuffix(s, origin)
+		} else {
+			d_origin := "." + origin
```
Comment (miekg) [inline dnsutil/util.go]: try not to use underscores in var names

File: dnsutil/util.go [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	if s == "." {
+		return "."
+	}
+
+	lod := origin[len(origin)-1] == '.' // Is the last char of origin a dot?
+	lsd := s[len(s)-1] == '.'           // Is the last char of s a dot?
+
+	if lod == lsd { // Either they both end in ".", or neither do.
+		if origin == s {
```
Comment (miekg) [inline dnsutil/util.go]: these if's might be better rewritten as a switch?

File: dnsutil/util_test.go [hunk 14]
```diff
@@ -0,0 +1,110 @@
+package dnsutil
+
+import "testing"
+
+// func AddDomain
+
+func TestAddDomain_nodot(t *testing.T) {
```
Comment (miekg) [inline dnsutil/util_test.go]: TestAddDomainNameNoDot

File: dnsutil/util_test.go [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+			t.Errorf("%v: expected (%v) got (%v)\n", test.experiment, test.expected, actual)
+		}
+	}
+}
+
+func TestAddDomain_dot(t *testing.T) {
+	var tests = []struct{ experiment, expected string }{
+		{"@", "example.com."},
+		{"foo", "foo.example.com."},
+		{"foo.", "foo."},
```
Comment (miekg) [inline dnsutil/util_test.go]: this repeats the same test as above...

File: dnsutil/util_test.go [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+				t.Errorf("TrimDomain(%#v, %#v): expected (%v) got (%v)\n", dom, test.experiment, test.expected, actual)
+			}
+		}
+	}
+}
+
+func TestTrimDomain_shortdomain(t *testing.T) {
+	// This test shouldn't be needed but I was weary of off-by-one errors
+	// due to a very short TLD.
+	var tests = []struct{ experiment, expected string }{
```
Comment (miekg) [inline dnsutil/util_test.go]: move tests that are actually different into TestStripDomainNameNormal (and drop Normal)

File: dnsutil/zone.go [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+type zoneGenData struct {
+	Origin     string
+	DefaultTtl uint32
+	Records    []dns.RR
+}
+
+func (z *zoneGenData) Len() int      { return len(z.Records) }
+func (z *zoneGenData) Swap(i, j int) { z.Records[i], z.Records[j] = z.Records[j], z.Records[i] }
+func (z *zoneGenData) Less(i, j int) bool {
+	a, b := z.Records[i], z.Records[j]
```
Comment (miekg) [inline dnsutil/zone.go]: to properly sort names you will need to resort to the wireformat, see https://github.com/miekg/dns/blob/dfefa36553955889dbb9ef892d4bd116ec095ce2/dnssec.go#L587

If think that something that *efficently* (is so far possible with go dns) sorts a list of RRs is useful

File: dnsutil/zone.go [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+	}
+	return a.String() < b.String()
+}
+
+func WriteZoneFile(w io.Writer, records []dns.RR, origin string, defaultTtl uint32) error {
+	z := &zoneGenData{
+		Origin:     origin,
+		DefaultTtl: defaultTtl,
+	}
+	for _, rec := range records {
```
Comment (miekg) [inline dnsutil/zone.go]: there shouldn't be a need to copy these per rr.

File: dnsutil/zone_test.go [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+
+	// Compare:
+	if buf2.String() != expected {
+		t.Fatalf("Regenerated zonefile does not match: got=(\n%v\n)\nexpected=(\n%v\n)\n", buf2.String(), expected)
+	}
+}
+
+// func WriteZoneFile
+
+func TestWriteZoneFile_simple(t *testing.T) {
```
Comment (miekg) [inline dnsutil/zone_test.go]: These might be better written as ExampleTests that check the output

File: dnsutil/zone.go [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+	}
+	if rrtypeA == dns.TypeMX {
+		ta2, tb2 := a.(*dns.MX), b.(*dns.MX)
+		pa, pb := ta2.Preference, tb2.Preference
+		return pa < pb
+	}
+	return a.String() < b.String()
+}
+
+func WriteZoneFile(w io.Writer, records []dns.RR, origin string, defaultTtl uint32) error {
```
Comment (miekg) [inline dnsutil/zone.go]: If we are going to add something like this, we should focus on making something that allows users to override the way a RR is printed. I.e. stuff like: omit class or ttl, only print the rdata (or a subset of the rdata). I.e. something super flexible. I have no good ideas (yet) on how something like that should look.

File: dnsutil/util.go (hunk 11 above)
Comment (TomOnTime) [inline dnsutil/util.go]: I'll leave that to another PR.

File: dnsutil/util.go (hunk 13 above)
Comment (TomOnTime) [inline dnsutil/util.go]: Refactored out of existence.

File: dnsutil/util.go (hunk 12 above)
Comment (TomOnTime) [inline dnsutil/util.go]: Done

File: dnsutil/util_test.go (hunk 15 above)
Comment (TomOnTime) [inline dnsutil/util_test.go]: Consolidated.

File: dnsutil/util_test.go (hunk 14 above)
Comment (TomOnTime) [inline dnsutil/util_test.go]: Agreed

File: dnsutil/util_test.go (hunk 16 above)
Comment (TomOnTime) [inline dnsutil/util_test.go]: Done

File: dnsutil/zone.go (hunk 18 above)
Comment (TomOnTime) [inline dnsutil/zone.go]: Fixed

File: dnsutil/zone.go (hunk 17 above)
Comment (TomOnTime) [inline dnsutil/zone.go]: I prefer this sort order because it groups hostnames in a way that makes it easier for people to see patterns: Grouping all records in a subdomain next to each other.  I see the value of having it sorted in DNSSEC order, but it isn't as readable, but the code is structured so it would be easy to abstract out the ordering code and create a WriteZoneFileOrdered() function that lets you supply a custom sort function.

File: dnsutil/util.go (hunk 10 above)
Comment (TomOnTime) [inline dnsutil/util.go]: Done

File: dnsutil/util.go (hunk 4 above)
Comment (TomOnTime) [inline dnsutil/util.go]: Ok

File: dnsutil/util.go (hunk 5 above)
Comment (TomOnTime) [inline dnsutil/util.go]: I'd rather leave it as is.  Often user-facing code displays FQDNs without the trailing dot.  Being able to include (or not include) the "." at the end of the origin makes it easy for this one function to work in either case.  That's why I didn't call this MakeFqdn() (and AppendIfNotFqdn() is a bit convoluted).

File: dnsutil/util.go (hunk 6 above)
Comment (TomOnTime) [inline dnsutil/util.go]: Done

File: dnsutil/util.go (hunk 7 above)
Comment (TomOnTime) [inline dnsutil/util.go]: Good catch.  It was needed for an earlier implementation but not any more. Removed. (actually... refactored out)

File: dnsutil/util.go (hunk 8 above)
Comment (TomOnTime) [inline dnsutil/util.go]: Done.

File: dnsutil/util.go (hunk 9 above)
Comment (TomOnTime) [inline dnsutil/util.go]: Done.

File: dnsutil/util.go (hunk 10 above)
Comment (asergeyev) [inline dnsutil/util.go]: I think you need to remove empty line between doc and package. You can check doc running ```godoc github.com/miekg/dns/dnsutil``` or use http-server of godoc (assuming you edit package inside GODOC).  Common practice to use "Package pkgname" as start of first sentence but godoc would work without it too.

File: dnsutil/zone.go (hunk 17 above)
Comment (asergeyev) [inline dnsutil/zone.go]: Maybe there should be WriteZoneFile should accept either constant for type of label ordering or comparator. I've encountered needs to have zones written in odd order before.  

Also I don't really know if it's useful but another way of writing zonefiles is by grouping things under $ORIGIN subsections of zone.

Comment (TomOnTime) [inline dnsutil/zone.go]: How about a "flag int" parameter that can be used in the future to supply flags for various sorting and formatting options?

File: dnsutil/util.go (hunk 10 above)
Comment (asergeyev) [inline dnsutil/util.go]: Looks great now!

File: dnsutil/util.go [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+// functions in package dns.
+package dnsutil
+
+import (
+	"strings"
+
+	"github.com/miekg/dns"
+)
+
+// AddDomain adds origin to s if s is not already a FQDN.
```
Comment (miekg) [inline dnsutil/util.go]: AddOrigin.

"The zonefile convention" is specified in RFC 1035, Section "5.1. Format", regarding the '@' sign.

File: dnsutil/util_test.go [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+		for i, test := range tests {
+			actual := TrimDomainName(test.experiment, dom)
+			if test.expected != actual {
+				t.Errorf("%d TrimDomainName(%#v, %#v): expected (%v) got (%v)\n", i, test.experiment, dom, test.expected, actual)
+			}
+		}
+	}
+}
+
+func TestTrimDomainNameShortSuffix(t *testing.T) {
```
Comment (miekg) [inline dnsutil/util_test.go]: Could you collopse this into 2 two test function for both AddOrigin and TrimDomainName

File: dnsutil/zone.go (hunk 17 above)
Comment (miekg) [inline dnsutil/zone.go]: Can we give it a sort.Sort like parameter that sorts for you?

Don't really like magical flags for this...

File: dnsutil/zone.go [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+	"fmt"
+	"io"
+	"log"
+	"sort"
+	"strings"
+
+	"github.com/miekg/dns"
+)
+
+type zoneGenData struct {
```
Comment (miekg) [inline dnsutil/zone.go]: Can you put this change in a separate PR? Still not sure I want it or not and in what form. We can then merge the other functions

File: dnsutil/util_test.go (hunk 22 above)
Comment (TomOnTime) [inline dnsutil/util_test.go]: Done

File: dnsutil/util.go (hunk 21 above)
Comment (TomOnTime) [inline dnsutil/util.go]: Added

File: dnsutil/zone.go (hunk 23 above)
Comment (TomOnTime) [inline dnsutil/zone.go]: I've removed it.  I'll make a new PR soon.

File: labels_test.go [hunk 24]
```diff
@@ -57,6 +58,8 @@ func TestSplit(t *testing.T) {
 			t.Logf("%s %v", s, Split(s))
 		}
 	}
+	f := Split(".www.miek.nl.")
+	fmt.Printf("TEST: %#v\n", f)
```
Comment (miekg) [inline labels_test.go]: what's this actually?

Fine with extending the test below.

File: labels.go [hunk 25]
```diff
@@ -102,6 +103,7 @@ func CountLabel(s string) (labels int) {
 
 // Split splits a name s into its label indexes.
 // www.miek.nl. returns []int{0, 4, 9}, www.miek.nl also returns []int{0, 4, 9}.
+// .www.miek.nl. returns []int{0, 1, 5, 10}
```
Comment (miekg) [inline labels.go]: hmmm, this might be a bug, because the domain name is not valid.... good also call it garbage in, garbage out.

File: labels_test.go (hunk 24 above)
Comment (TomOnTime) [inline labels_test.go]: Oops. That was supposed to be deleted before I pushed.  Fixed.

**Comment by miekg** [comment]
Now there are conflicts... could you fix as well, then I press the glorious merge button.

**Comment by TomOnTime** [comment]
I squashed the commits.
